src/lib/download-progress.ts

```diff
@@ -0,0 +1,88 @@
+import * as _ from 'lodash';
+import blink = require('./blink');
+import { DeviceStatus, TargetState } from '../types/state';
+import * as deviceState from '../device-state';
+import * as targetState from '../device-state/target-state';
+
+interface ServiceState {
+	id: string;
+	status: string;
+	download_progress: number | null;
+}
+
+const DOWNLOAD_BLINK_THROTTLE = 10000;
+const DOWNLOAD_BLINK_PERIOD_BASE = 100;
+const DOWNLOAD_BLINK_PERIOD_STEP = 10;
+
+const throttledBlink = _.throttle((progress: number) => {
```

Comment:
I wonder if because of the throttle might get to a case where the blinking gets stuck on a pattern. For instance, let's say the function gets called with `downloadProgress(95)`, and downloading the last 5% takes less than 10 seconds. The function will miss the `downloadProgress(100)` just as the download is finished and it will have to wait until the next time current state changes, which may be an undetermined span of time. 

It's probably not that bad, but it may lead to ambiguous signaling.

Reply:
Yes that's certainly a possibility and it's the reason the throttle is 10s and not more. From my (limited) experience even when reaching 100% the supervisor takes way longer to apply the update. During that time `update_pending` is still set to `true` so the throttled blink function will still get invoked and turn off the LED. 

As an extra precaution I also added [this](https://github.com/balena-os/balena-supervisor/blob/aa2370689491abcffa25a1e1c7a2c9f0cbd91241/src/lib/download-progress.ts#L58) which should always be triggered when `update_pending` goes from `true` to `false` (according to my understanding).

So my conclusion is that yes you could be reporting ambiguous signalling but most probably for at most 10 seconds.